appscale/tools/agents/ec2_agent.py

```diff
@@ -796,12 +905,22 @@ def cleanup_state(self, parameters):
 
     AppScaleLogger.log("Deleting security group {0}".format(
       parameters[self.PARAM_GROUP]))
+    retries_left = self.SECURITY_GROUP_RETRY_COUNT
     while True:
       try:
-        conn.delete_security_group(parameters[self.PARAM_GROUP])
+        sg = self.get_security_group_by_name(parameters,
+                                             parameters[self.PARAM_GROUP])
+        conn.delete_security_group(group_id=sg.id)
         return
       except EC2ResponseError:
-        time.sleep(5)
+        time.sleep(self.SLEEP_TIME)
+        retries_left -= 1
+      except AgentRuntimeException:
+        AppScaleLogger.log('Could not find security group {}, delete '
+                           'successful.'.format(parameters[self.PARAM_GROUP]))
+        if retries_left == 0:
```

Comment:
Shouldn't this be in the other except clause?

Reply:
```suggestion
  def get_security_group_by_name(self, parameters, group):
    """Gets a security group in AWS with the given name.
     Args:
      parameters: A dict that contains the credentials necessary to authenticate
        with AWS.
      group: A str that names the group that should be found.
    Returns:
      The 'boto.ec2.securitygroup.SecurityGroup' that has the correct group
      name.
    Raises:
      EC2ResponseError: If there was an error contacting EC2.
      AgentRuntimeException: If the security group could not be found.
    """
    conn = self.open_connection(parameters)
    try:
      return next(sg for sg in conn.get_all_security_groups()
                  if sg.name == group)
    except StopIteration:
      raise AgentRuntimeException('Could not find security group with name '
                                  '{}!'.format(group))

```

and then in 
```suggestion
      except EC2ResponseError:
        retries_left -= 1
        if retries_left == 0:
          break
        time.sleep(self.SLEEP_TIME)
        
```
and leave the AgentRuntimeException one to return if it couldn't find it?

and also catch the EC2ResponseError in other methods that call get_security_group_by_name

Reply:
Looks like we edited our comments at roughly the same time :smile: 

I think what you proposed is a step in the right direction. You probably want to `raise` instead of `break` when the retries are up, though. I also still think `AgentRuntimeException` shouldn't be used to indicate that the SG does not exist since it is used in other places to indicate a failure or transient error.